### ISSUE TITLE
how to achieve nested filters query?

### ISSUE THREAD

**Issue by anjnaharsora** [automategreen/dynamoose#141]
Friends.scan({
        or: [
            {and: [{userRef: {eq: {id: param.user}}, friend: {eq: {id: param.friend}}}]},
            {and:[{userRef: {eq: {id: param.friend}}, friend: {eq: {id: param.user}}}]}
        ]
    }
Example.

**Comment by brandongoode** [comment]
Currently, this isn't supported in Dynamooose.  While DynamoDB does support this logic, it requires the use of a FilterExpression:

http://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Scan.html#DDB-Scan-request-FilterExpression

Dynamoose uses the legacy ScanFilter which doesn't support nesting:

http://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Scan.html#DDB-Scan-request-ScanFilter


This is something I would like to implement int he future, but haven't had the time to do.

See #65

**Comment by anjnaharsora** [comment]
thanks @brandongoode

**Comment by lrcry** [comment]
Hi all

I think it would be a nice idea for now that dynamoose allows raw filter in scan i.e. `Model.scan(filter, options, callback)` where a new option `useRawAwsFilter` could be added, so that we could use what is currently supported in aws-sdk.

```
const filter = {
  TableName: 'TestTable',
  FilterExpression: 'test.title = :title',
  ExpressionAttributeValues: {
    ':title': 'dynamotest'
  }
};
TestTable.scan(filter,{ 'useRawAwsFilter': true }).exec().then(...).catch(...)
```

I would like to try to fork one to see how it goes.

Regards

Hans

**Comment by brandongoode** [comment]
@lrcry That great.  I was thinking about this some over the past week and wanted to share my thoughts.  

I don't think the `useRawAwsFilter` flag is needed. We could just assume if the `filterExpression` option is present then it should take priority.  We need to make sure to account for attribute names.  Also, in addition to passing in a filter option, I think having the scan building function too would be great.

Something like this:

```js
GameResults
  .scan()
  .filterExpression('#s > :score')
  .expressionAttributeValues({ score : 100 })
  .expressionAttributeNames({ s : 'score'})
```

or

```js
GameResults
  .scan({
    filterExpression: '#s > :score',
    expressionAttributeValues: { score : 100 },
    expressionAttributeNames: { s : 'score'}
  })
```

**Comment by lrcry** [comment]
@brandongoode  This syntax you suggested seems similar to [this](https://github.com/clarkie/dynogels#scan). BTW I got the one I raised implemented just now.

Just wondering if it is enough for now as I can see that quite a few people need the nested scan.

**Comment by brandongoode** [comment]
Yes, it is very similar.  A few minor differences around name and value definition.  My goal is to have it align the current implementation of conditions (really just filter expressions) in `.save` - https://dynamoosejs.com/api#modelputoptions-callback--modelsaveoptions-callback. 

I try to make sure all query and scans can be build using and object or the helper function. 

Please go ahead and submit the PR for review.  The functions can be added later.

**Comment by jderose9** [comment]
@lrcry I'm looking forward to this... thanks for the effort!

**Comment by brandongoode** [comment]
@lrcry Do you know when you plan to have this PR ready.  I'm planning to release 0.8 sometime over the next week.  I would like to include this in that release if you think it will be ready.

**Comment by lrcry** [comment]
@brandongoode @jderose9 I'm still tackling with some problem over there. Trying to get it in v0.8. Thanks

**Comment by lrcry** [comment]
[#158](https://github.com/automategreen/dynamoose/pull/158/)

I've opened a PR here. Please feel free to tell me your thoughts.

Cheers ;)

**Comment by brandongoode** [comment]
Closes by #158
